client/src/components/graph/overlays/centroidLabels.js

```diff
@@ -15,27 +16,74 @@ export default
   showLabels: state.centroidLabels?.showLabels,
 }))
 class CentroidLabels extends PureComponent {
+  static watchAsync(props, prevProps) {
+    return !shallowEqual(props.watchProps, prevProps.watchProps);
+  }
+
   constructor(props) {
     super(props);
     this.state = {
-      status: "pending",
-      labels: new Map(),
-      colorAccessor: null,
+      labelSize: 0,
     };
   }
 
-  componentDidMount() {
-    this.updateState(null);
-  }
-
-  componentDidUpdate(prevProps) {
-    this.updateState(prevProps);
+  componentDidUpdate() {
+    // this.updateState(prevProps);
 
     const { showLabels, overlaySetShowing } = this.props;
-    const { labels } = this.state;
-    overlaySetShowing("centroidLabels", showLabels && labels.size > 0);
+    const { labelSize } = this.state;
+    overlaySetShowing("centroidLabels", showLabels && labelSize > 0);
   }
 
+  fetchAsyncProps = async (props) => {
+    const {
+      annoMatrix,
+      colors,
+      layoutChoice,
+      categoricalSelection,
+    } = props.watchProps;
+    const { schema } = annoMatrix;
+    const { colorAccessor } = colors;
+
+    const [layoutDf, colorDf] = await this.fetchData();
+    let labels = [];
+    if (colorDf) {
+      labels = calcCentroid(
+        schema,
+        colorAccessor,
+        colorDf,
+        layoutChoice,
+        layoutDf
+      );
+    }
+
+    this.setState({ labelSize: labels.size });
```

Comment:
Why not execute `overlaySetShowing()` here instead of adding `labelSize` to state,  re-rendering, and then calling `overlaySetShowing()` in `componentDidUpdate()`?

Reply:
interesting idea - will explore it.

Reply:
Made the proposed change.